### PR TITLE
Fix bug of giveall command

### LINKED ISSUE #166: [PE-D] Users are able to add duplicate assignments (giveall)
And due dates are not changed.

To reproduce:
1. Have a number of students already with that assignment (i.e., Lab3) in the list (but with a different due date)

2. giveall m/cs1231s d/Lab3 by/ 14/06/2015

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: leezhixuan/ped#11

### LINKED ISSUE #180: [PE-D] Duplicated Assignments
Steps:

giveall m/CS2100 d/Assignment 2 by/ 15/10/2021,1300

giveall m/CS2100 d/Assignment 2 by/ 15/10/2021,1300

Expected:

Successfully given 2 copies of Assignment 2, or error message

Actual:

New assignment added to all persons in CS2100: Assignment 2, DueDate: 15 Oct 2021, 01:00 PM, Status: [PENDING], but only 1 copy of Assignment2 found

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: guangxuan/ped#9

### PR DESCRIPTION
**Pull request by TaoJun99**
Fixes #166 
Fixes #180 

Add error message when trying to give assignment to all students but all students already have the specified assignment

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/AddAssignmentToAllCommand.java [hunk 1]
```diff
@@ -50,17 +56,42 @@ public AddAssignmentToAllCommand(Module module, Assignment assignment) {
...
+                .stream()
+                .collect(Collectors.partitioningBy(person -> model.hasAssignment(person, toAdd))).values());
+
+        List<Person> personListWithoutAssignment = filteredPersonList.get(0);
+        List<Person> personListWithAssignment = filteredPersonList.get(1);
+
+        Assignment existingAssignment = toAdd;
+        if (!personListWithAssignment.isEmpty()) {
+            existingAssignment = personListWithAssignment.get(0).getAssignments().getAssignment(toAdd.getDescription());
+            if (!existingAssignment.getDueDate().equals(toAdd.getDueDate())) {
```
Comment (Droffilc13) [inline src/main/java/seedu/address/logic/commands/AddAssignmentToAllCommand.java:80]: create a method Assignment#isSameDueDate instead. potentially violates law of demeter

File: src/main/java/seedu/address/logic/commands/AddAssignmentToAllCommand.java [hunk 2]
```diff
@@ -50,17 +56,42 @@ public AddAssignmentToAllCommand(Module module, Assignment assignment) {
...
+        //Split the persons into those with or without the assignment
+        List<List<Person>> filteredPersonList = new ArrayList<>(personListInModule
+                .stream()
+                .collect(Collectors.partitioningBy(person -> model.hasAssignment(person, toAdd))).values());
+
+        List<Person> personListWithoutAssignment = filteredPersonList.get(0);
+        List<Person> personListWithAssignment = filteredPersonList.get(1);
+
+        Assignment existingAssignment = toAdd;
+        if (!personListWithAssignment.isEmpty()) {
```
Comment (Droffilc13) [inline src/main/java/seedu/address/logic/commands/AddAssignmentToAllCommand.java:78]: Perhaps wrap all this code up in a method and give it a suitable name? Method a bit long

**Review by Droffilc13 — CHANGES_REQUESTED** [review]
Test case, I trust you.
 
There are some changes I have suggested for your consideration

File: src/main/java/seedu/address/logic/commands/AddAssignmentToAllCommand.java [hunk 3]
```diff
@@ -50,17 +56,42 @@ public AddAssignmentToAllCommand(Module module, Assignment assignment) {
...
+            }
+        }
+
+        if (personListWithoutAssignment.isEmpty()) {
+            throw new CommandException(String.format(MESSAGE_ALL_HAS_ASSIGNMENT, module));
+        }
+
+        // Create a new assignment with the same description as existing ones
+        // to prevent inconsistencies in letter cases
+        Assignment standardisedAssignment = new Assignment(existingAssignment.getDescription(),
```
[the remark this answers is not in the document]
Comment (TaoJun99) [reply src/main/java/seedu/address/logic/commands/AddAssignmentToAllCommand.java:92]: In this case, I am using description from existingAssignment and the other two fields from toAdd, so I am not copying the whole assignment over

File: src/main/java/seedu/address/logic/commands/AddAssignmentToAllCommand.java (hunk 1 above)
Comment (TaoJun99) [reply src/main/java/seedu/address/logic/commands/AddAssignmentToAllCommand.java:80]: Will do!

File: src/main/java/seedu/address/logic/commands/AddAssignmentToAllCommand.java (hunk 2 above)
Comment (TaoJun99) [reply src/main/java/seedu/address/logic/commands/AddAssignmentToAllCommand.java:78]: Will do!

File: docs/UserGuide.md [hunk 4]
```diff
@@ -216,9 +216,20 @@ Adds an assignment with a deadline to all students in the specified module .
 
 Format: `giveall m/MODULE d/DESCRIPTION by/ d/M/YYYY [,HHmm]`
 
+**Note:**
+1. It is possible to add an assignment using `giveall` even if some students have the assignment already. Students who have the 
+assignment will not receive a duplicate assignment.
```
Comment (Droffilc13) [inline docs/UserGuide.md:221]: does this space affect visuals

File: docs/UserGuide.md [hunk 5]
```diff
@@ -216,9 +216,20 @@ Adds an assignment with a deadline to all students in the specified module .
...
+2. When using `giveall` in the situation stated in point 1, the specified assignment in the input needs to have the same 
+due date as the existing assignment. **The command will not be valid if the due date is different.** 
+Point 3 of the examples section illustrates this.
```
Comment (Droffilc13) [inline docs/UserGuide.md:224]: There is no numbered points in the user guide for examples. Possible to omit the "point x" and say `If some students have the same assignments you wish to give but of a different due date, the command will not work to ensure consistency."?

If you take my suggestion, maybe can put this point in a note box

File: src/main/java/seedu/address/logic/commands/AddAssignmentToAllCommand.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Droffilc13) [reply src/main/java/seedu/address/logic/commands/AddAssignmentToAllCommand.java:85]: i see

File: src/main/java/seedu/address/logic/commands/AddAssignmentToAllCommand.java [hunk 6]
```diff
@@ -50,19 +56,60 @@ public AddAssignmentToAllCommand(Module module, Assignment assignment) {
...
+        //Split the persons into those with or without the assignment
+        List<List<Person>> filteredPersonList = new ArrayList<>(personListInModule
+                .stream()
+                .collect(Collectors.partitioningBy(person -> model.hasAssignment(person, toAdd))).values());
 
-        return new CommandResult(String.format(MESSAGE_SUCCESS, module, toAdd));
+        List<Person> personListWithoutAssignment = filteredPersonList.get(0);
+        List<Person> personListWithAssignment = filteredPersonList.get(1);
+
+        Assignment existingAssignment = getExistingAssignment(personListWithAssignment);
```
Comment (Droffilc13) [inline src/main/java/seedu/address/logic/commands/AddAssignmentToAllCommand.java:77]: I'm not sure how to suggest, but the name to me is not really matching? but then again, idk what a more appropriate name is

**Review by Droffilc13 — CHANGES_REQUESTED** [review]

**Review by Droffilc13 — APPROVED** [review]
LGTM

**Pull request merged**